src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -473,5 +496,143 @@ private ImmutableArray<AsyncCompletionData.CompletionFilter> GetFilters(RoslynCo
 
             return current == questionPosition;
         }
+
+        /// <summary>
+        /// Provides an efficient way to compute a set of completion filters associated with a collection of completion items.
+        /// Presence of expander and filter in the set have different meanings. Set contains a filter means the filter is
+        /// available but unselected, whereas it means available and selected for an expander. Note that even though VS support 
+        /// having multiple expanders, we only support one.
+        /// </summary>
+        private class FilterSet
+        {
+            // Cache all the VS completion filters which essentially make them singletons.
+            // Because all items that should be filtered using the same filter button must 
+            // use the same reference to the instance of CompletionFilter.
+            private static readonly Dictionary<string, AsyncCompletionData.CompletionFilter> s_filterCache =
+                new Dictionary<string, AsyncCompletionData.CompletionFilter>();
+
+            private static readonly ImmutableArray<int> s_filterMasks;
+            private static readonly int s_expanderMask;
```

Comment:
>static [](start = 20, length = 6)

why static?
Should this be the same for C#/VB?

same for statics above

Reply:
Static so those masks only needs to be initialed once, in the static ctor (they never change). Not sure I understand the comment on C#/VB. Could you please explain?

Reply:
Let us assume within the same VS session we are called for different languages. What would happen for s_expanderMask? Should it be specific for each language?
Would it grow more and more?

---
In reply to: [296019303](https://github.com/dotnet/roslyn/pull/36146#discussion_r296019303) [](ancestors = 296019303)

Reply:
It should be fine. The mask is created based on[ `CompletionItemFilter.AllFilters`](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionItemFilter.cs,281ef61b5aab46ed,references), which is language agnostic.